AST-based query fuzzing mode for clickhouse-client

akuzm opened the pull request:
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement AST-based query fuzzing mode for clickhouse-client. See [this label](https://github.com/ClickHouse/ClickHouse/issues?q=label%3Afuzz+is%3Aissue) for the list of issues we recently found by fuzzing. Most of them were found by this tool, and a couple by SQLancer and `00746_sql_fuzzy.pl`.

akuzm committed:
Fuzzer

akuzm committed:
streaming wip

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
more fuzzing

alexey-milovidov commented on programs/client/QueryFuzzer.h:
```diff
@@ -0,0 +1,49 @@
+struct QueryFuzzer
+{
+    //pcg64 fuzz_rand{static_cast<UInt64>(rand())};
+    pcg64 fuzz_rand{clock_gettime_ns()};
```
There is `randomSeed` function.

alexey-milovidov commented on programs/client/Client.cpp:
```diff
@@ -2060,13 +2280,58 @@ class Client : public Poco::Util::Application
+    const ucontext_t signal_context = *reinterpret_cast<ucontext_t *>(context);
+    const StackTrace stack_trace(signal_context);
+    std::cerr << fmt::format("Received signal {} at: {}", sig,
+        stack_trace.toString()) << std::endl;
```
toString is not signal-safe. std::cerr and fmt::format neither.

akuzm replied:
I want to have stack traces when the client dies, but don't want to have a separate thread in client to handle these faults (that's complicated). This is for deadly signals, the app is going to terminate anyway, so maybe it's OK to use not signal-safe functions?

akuzm replied:
Discussed on zoom -- we had that before and it lead to deadlocks. The solution is to copy the handling we have in BaseDaemon.

akuzm replied:
I removed it for now because I want to merge soon and run the fuzzer in CI. I'll add proper signal handling separately.

akuzm committed:
fixes

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
cleanup

akuzm committed:
cleanup

akuzm committed:
style

alexey-milovidov commented:
Please link all the issues found by this fuzzer to let people know how good this tool is.

akuzm committed:
fix builds

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm commented:
> Please link all the issues found by this fuzzer to let people know how good this tool is.

Added the list of issues to the changelog entry.

The tests are green and I cleaned up the code, so merging now. Next step is to start running it in the CI.

alexey-milovidov commented on programs/client/Client.cpp:
```diff
@@ -762,17 +774,137 @@ class Client : public Poco::Util::Application
+                std::cerr << "Read too much text and still can't parse a query."
+                     " Aborting." << std::endl;
+                last_exception_received_from_server
+                    = std::make_unique<Exception>(1, "~");
```
This will require cleanups.

alexey-milovidov commented on programs/client/Client.cpp:
```diff
@@ -871,6 +1003,121 @@ class Client : public Poco::Util::Application
+                    // Debug AST cloning errors.
+                    if (base_before_fuzz != base_after_fuzz)
+                    {
+                        fprintf(stderr, "base before fuzz: %s\n"
```
It represents you as a seasoned C developer.

alexey-milovidov commented on programs/client/Client.cpp:
```diff
@@ -1895,6 +2146,7 @@ class Client : public Poco::Util::Application
             ("highlight", po::value<bool>()->default_value(true), "enable or disable basic syntax highlight in interactive command line")
             ("log-level", po::value<std::string>(), "client log level")
             ("server_logs_file", po::value<std::string>(), "put server logs into specified file")
+            ("query-fuzzer-runs", po::value<int>()->default_value(0), "query fuzzer runs")
```
Minimal comment about usage is required.
